src/python/pants/backend/python/target_types_rules.py

```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
     inject_for = PythonDistributionDependencies
 
 
 @rule
 async def inject_python_distribution_dependencies(
-    request: InjectPythonDistributionDependencies,
+    request: InjectPythonDistributionDependencies, python_infer_subsystem: PythonInferSubsystem
 ) -> InjectedDependencies:
-    """Inject any `.with_binaries()` values, as it would be redundant to have to include in the
-    `dependencies` field."""
-    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
-    with_binaries = original_tgt.target[PythonProvidesField].value.binaries
-    if not with_binaries:
+    """Inject dependencies that we can infer from entry points in the distribution."""
+    if not python_infer_subsystem.entry_points:
         return InjectedDependencies()
-    # Note that we don't validate that these are all `pex_binary` targets; we don't care about
-    # that here. `setup_py.py` will do that validation.
-    addresses = await Get(
-        Addresses,
-        UnparsedAddressInputs(
-            with_binaries.values(), owning_address=request.dependencies_field.address
+
+    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
+    explicitly_provided_deps, entry_points = await MultiGet(
+        Get(ExplicitlyProvidedDependencies, DependenciesRequest(original_tgt.target[Dependencies])),
+        Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(
+                original_tgt.target[PythonDistributionEntryPoints]
+            ),
         ),
     )
-    return InjectedDependencies(addresses)
+
+    entry_point_addresses = Addresses()
+
+    if entry_points.val:
+        address = original_tgt.target.address
+        all_entry_points = [
+            (key, name, entry_point)
+            for key, section in entry_points.val.items()
+            for name, entry_point in section.items()
+        ]
+        all_owners = await MultiGet(
+            Get(PythonModuleOwners, PythonModule(entry_point.module))
+            for _, _, entry_point in all_entry_points
+        )
+        for (key, name, entry_point), owners in zip(all_entry_points, all_owners):
+            field_str = repr({key: {name: entry_point.spec}})
+            explicitly_provided_deps.maybe_warn_of_ambiguous_dependency_inference(
+                owners.ambiguous,
+                address,
+                import_reference="module",
+                context=(
+                    f"The distribution target {address} has the field "
+                    f"`entry_points={field_str}`"
+                ),
+            )
+            maybe_disambiguated = explicitly_provided_deps.disambiguated(owners.ambiguous)
+            unambiguous_owners = Addresses(
+                owners.unambiguous
+                or ((maybe_disambiguated,) if maybe_disambiguated else Addresses())
+            )
+            entry_point_addresses += unambiguous_owners  # type: ignore[assignment]
```

Comment:
I don't think this will work. `Addresses` is immutable. Instead, you need to reassign it.

Reply:
hehe, maybe it is my C background shining through. But this is reassignment. `a += b` translates to `a = a + b`. ;)

Reply:
But lets ensure there is a test, that proves it...

Reply:
Also, MyPy agrees...
```
src/python/pants/backend/python/target_types_rules.py:307:13: error:
Incompatible types in assignment (expression has type "Tuple[Address, ...]",
variable has type "Addresses")  [assignment]
```

Reply:
Hm, to avoid confusion, how about using `module_owners = OrderedSet()` at the start of this `for` loop, then calling `.update()`(?) on it each iteration. After the loop, set `entry_point_addresses` equal to `entry_points.pex_binary_addresses` + `module_owners`.